blog/06-02-2023-catering-to-the-client-side.md

```diff
@@ -0,0 +1,173 @@
+---
+slug: "catering-to-the-client-side"
+title: "Catering to the client-side"
+date: 2023-02-06
+authors: [petehodgson]
+description: "TODO"
+draft: false
+---
+
+While OpenFeature initially focused on support for server-side feature flagging, we know that a lot of feature-flagging (likely the majority) happens on the client - mobile apps and frontend web apps. As such, we're currently finalizing [a proposal](https://github.com/open-feature/spec/issues/167) which extends the OpenFeature spec to support client-side use cases. 
+
+In this post I'll summarize those changes, but to understand them in context we'll first talk about what makes client-side feature flagging different before diving into how that will impact the OpenFeature APIs.
+
+## Context is King: why client-side flagging is different
+
+### Static evaluation context
+
+The biggest difference between the client- and server-side feature flagging is the relative frequency of evaluation context changes within those two paradigms. In the typical server-side scenario - a service responding to http requests - the context for a feature flagging decision changes with every request. Every new request is coming from a new user, and a lot of the evaluation context which affects a feature flagging decision is based on the user making the request.
+
+In contrast, with a client-side app all feature flagging decisions are made in the context of the same user - the user interacting with the client-side app - and so the evaluation context is quite static. There are cases where evaluation context will change within a client-side app - when a user logs in, for example - but by and large with client-side code we can treat feature flag evaluation context as something that is fixed (while still providing mechanisms to update it).
+
+### Flag evaluation is slow
+
+With server-side flags, we can assume that evaluating a feature flag is a relatively fast operation. In some scenarios the flagging rules can be evaluated in-process, and if not then flag evaluation is typically a quick service call - akin to making a DB query or calling a redis cache.
+
+This situation is quite different with client-side flags. It's typically not possible to evaluate a flagging decision locally. Local evaluation would require the full feature flagging ruleset to be exposed to the client, something many organizations are not comfortable with. Additionally, flag evaluation may require additional context that's simply not available on the client at the time of evalution. 
```

Comment:
> Local evaluation would require the full feature flagging ruleset to be exposed to the client, something many organizations are not comfortable with.

This is the way some vendors do it though (I think 2 we know of). I think it's valid, though there are tradeoffs. One advantage of doing things this way is little (or possible no) PII leaves the client app for evaluation purposes. As we've mentioned before, SDKs built this way can conform to our proposed paradigm, while the reverse isn't true.

I'm not sure how valuable it is to mention this, but I would at the least, suggest that both patterns are valid.

Reply:
I agree that this model exists, but AFAICT it's pretty rare. Given that, I'd rather not distract the reader.